### PR TITLE
Integrate currencies smart setting mechanism using third party API

### PR DESCRIPTION
**Pull request by ysv**

### CODE REVIEW

File: config/seed-v2/datasource.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+datasources:
```
Comment (mod) [inline config/seed-v2/datasource.yml:1]: config/seed.yml

File: config/seed-v2/datasource.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  - adapter: coingecko
```
Comment (ysv) [inline config/seed-v2/datasource.yml:2]: `seed:currencies:prefetech` (try to find better name)


```ruby
class Base
  def initialize(settings)
   @settings = settings
  end

  def prefetch_currencies
    # No implementation
  end
end

module Seed 
  class Coingecko < Base
    def prefetch_currencies
       # logic
    end
  end
end
```

rake
```ruby
YAML.load()...
"Seed::#{config[:adapter].camelize}".constantize => Seed::Coingecko
```

File: config/seed-v2/datasource.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    tarrifs:
+      # Defined in base currency
+      deposit_fee:            1
+      min_deposit_amount:     1
+      min_collection_amount:  1
+      withdrawaw_fee:         1
+      min_withdraw_amount:    1
+      withdraw_limit_24h:     1
+      withdraw_limit_72h:     3
+
```
Comment (ysv) [inline config/seed-v2/datasource.yml:33]: etherscan:
  endpoint: api
  api_key: 123API

File: app/services/prefill_service.rb [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    @tarrifs = datasource[:tarrifs]
+  end
+
+  # TODO: Rate limiting, Logger info, improve adapter logic
+  def perform
+    binding.pry
+    seeds = []
+    global_info = @adapter.get("/search", locale: "en")
+    mandatory_coins = global_info[:coins].select { |coin| coin[:id].in? @filters[:coingecko_ids] }
+    platform_coins = global_info[:coins].reject { |coin| coin[:market_cap_rank].nil? || coin[:market_cap_rank] > @filters[:market_cap_rank] }
```
Comment (ysv) [inline app/services/prefill_service.rb]: use select

File: app/services/prefill_service.rb [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      token_info = @adapter.get("/coins/#{coin[:id]}", COINGECKO_INFO_PARAMS)
+      seeds << form_payload(token_info)
+    end
+
+    platform_coins.each do | coin|
+      token_info = @adapter.get("/coins/#{coin[:id]}", COINGECKO_INFO_PARAMS)
+      seeds << form_payload(token_info) if token_info[:asset_platform_id].in? @filters[:asset_platform_ids]
+    end
+
+    File.open(Rails.root.join("config/seed/#{@output_file}"), "w") { |f| f.write seeds.to_yaml }
```
Comment (ysv) [inline app/services/prefill_service.rb]: YAML.write

File: app/services/prefill_service.rb [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+  end
+
+  def form_payload(info)
+    current_price = info.dig(:market_data, :current_price, @base_currency.to_sym).to_d
+    raw_precision = Math.log(current_price * 100, 10)
```
Comment (ysv) [inline app/services/prefill_service.rb]: move precision into separate method

File: app/services/prefill_service.rb [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+          withdraw_fee: @tarrifs[:withdraw_fee] / current_price,
+          min_withdraw_amount: @tarrifs[:min_withdraw_amount] / current_price,
+          withdraw_limit_24h: @tarrifs[:withdraw_limit_24h] / current_price,
+          withdraw_limit_72h: @tarrifs[:withdraw_limit_72h] / current_price
+
+      }
+      payload = {
+          code:           info[:symbol],
+          prefill_id:     info[:name],
+          coingecko_id:   info[:id],
```
Comment (ysv) [inline app/services/prefill_service.rb]: save coingecko_id and all specific attrs into options

File: app/services/prefill_service.rb [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                  0
+                else
+                  raw_precision.ceil + 1
+                end
+    if info[:asset_platform_id] == "ethereum"
+      etherscan_data = etherscan_contract_info(info[:contract_address])
+      if etherscan_data.present? && etherscan_data[:status] == 429
+        return {payload: {code: info[:symbol], status: 429}, category: :erc20}
+      end
+      currency = {
```
Comment (ysv) [inline app/services/prefill_service.rb]: build base payload then switch case for each asset_platform and add additional data

File: app/services/prefill_service.rb (hunk 7 above)
Comment (ysv) [reply app/services/prefill_service.rb]: log message for unsupported asset platfrom

File: config/seed-v2/datasource.yml (hunk 2 above)
Comment (dpatsora) [reply config/seed-v2/datasource.yml:2]: seed:prefetch:currencies

File: app/services/prefill_service.rb (hunk 5 above)
Comment (dpatsora) [reply app/services/prefill_service.rb]: I decided to use 
```ruby
File.open( output_file, 'w') { |f| f.write seeds.each(&:deep_stringify_keys!).to_yaml }
```

**Pull request closed without merging**